drbg/src/hashdrbg.rs

```diff
@@ -0,0 +1,220 @@
+        let no_of_bits_to_return: &[u8] = &(Hash::SeedLenBytes::U32 * 8).to_be_bytes();
+        for i in 0..len {
+            let counter_byte: &[u8] = &[counter];
+            let hashables = once(&counter_byte)
+                .chain(once(&no_of_bits_to_return))
+                .chain(data.iter());
+            let out = Self::hash(hashables);
+            tmp[i * Hash::OutputSize::USIZE..(i + 1) * Hash::OutputSize::USIZE]
+                .copy_from_slice(&out);
+            counter += 1;
+        }
+        let remainder = Hash::SeedLenBytes::USIZE % Hash::OutputSize::USIZE;
+        // we may need to do the hashing on last time
+        if remainder != 0 {
+            let counter_byte: &[u8] = &[counter];
+            let hashables = once(&counter_byte)
+                .chain(once(&no_of_bits_to_return))
+                .chain(data.iter());
+            let out = Self::hash(hashables);
+            tmp[len * Hash::OutputSize::USIZE..].copy_from_slice(&out[..remainder]);
+        }
+        tmp
+    }
+    /// This is the `Hash_DRBG_Instantiate_algorithm()` from _10.1.1.2_.
+    ///
+    /// Note that security strength is not used in the
+    /// implementation so it is not passed here.
+    fn instantiate_inner(
+        entropy_input: &[u8],
+        nonce: &[u8],
+        personalization_string: &[u8],
+    ) -> Self {
+        let v = Self::hash_df(&[entropy_input, nonce, personalization_string]);
+        let c = Self::hash_df(&[&[0], v.as_slice()]);
+        Self {
+            v,
+            c,
+            reseed_counter: 1,
+        }
+    }
+    /// This is the `Hash_DRBG_Reseed_algorithm()` from _10.1.1.3_.
+    fn reseed_inner(&mut self, entropy_input: &[u8], additional_input: &[u8]) {
+        self.v = Self::hash_df(&[&[0x1], self.v.as_slice(), entropy_input, additional_input]);
+        self.c = Self::hash_df(&[&[0x0], self.v.as_slice()]);
+        self.reseed_counter = 1;
+    }
+    /// This is the `Hash_DRBG_Generate_algorithm()` from _10.1.1.4_.
+    fn generate_inner(&mut self, additional_input: &[u8], out: &mut [u8]) {
+        let p = U896::from_u64(2).shl(Hash::SeedLenBytes::USIZE * 8);
+        let mut v = slice_to_u896(&self.v);
+        if !additional_input.is_empty() {
+            let w = Self::hash([&[0x2], self.v.as_slice(), additional_input].iter());
+            // TODO U896 is to much for most of the drbgs
+            // try to optimize later by using exactly U440 or U888
+            let w = slice_to_u896(&w);
+            v = v.add_mod(&w, &p);
+            self.v = truncate_from_start(&v.to_be_byte_array());
+        }
+        self.hash_gen(out);
+        let h = Self::hash([&[0x3], self.v.as_slice()].iter());
+        let h = slice_to_u896(&h);
+        let c = slice_to_u896(&self.c);
+        let reseed_counter = U896::from_u64(self.reseed_counter);
+        // we do V = (V + H + C + reseed_counter) mod 2^seedlen in multiple steps.
+        let v = v.add_mod(&h, &p);
```

Comment:
When we are doing modulo 1 << (x*8), we are basically just throwing most significant bytes away, leaving the x least significant bytes in the bignum. There is a chance we can optimize:
- if the BigNum is created to hold only (x+7)/8 limbs (u32/u64)
- the BigNum is allowed to overflow during add.
Then we just add without modulo, and return the x least significant bytes. But perhaps a future optimization as you mention above - lets leave as is for now.